### PR TITLE
FDTD solver

### PR DESCRIPTION
**Pull request by anekslen**
Add Standard and Non-Standard Finite Difference Time Domain Solvers to IPPL.

This replaces the Solver part of pullrequest #289.

There are two testcases for both the solvers in test/solvers/. One of them is the convergence test. The results for both solvers can be seen in the images below:

NonStandardFDTDSolver_convergence
StandardFDTDSolver_convergence

### CODE REVIEW

File: src/MaxwellSolvers/FDTDSolverBase.hpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    /**
+     * @brief Shifts the time for the fields.
+     * 
+     * Copies the current field values to the previous time step field and the next time step field values to the current field.
+     */
+    template <typename EMField, typename SourceField, fdtd_bc boundary_conditions>
+    void FDTDSolverBase<EMField, SourceField, boundary_conditions>::timeShift() {
+        // TODO: should this comment stay?
```
Comment (aaadelmann) [inline src/MaxwellSolvers/FDTDSolverBase.hpp:65]: yes the comments can stay

File: src/MaxwellSolvers/FDTDSolverBase.hpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        auto Eview  = Maxwell<EMField, SourceField>::En_mp->getView();
+        auto Bview  = Maxwell<EMField, SourceField>::Bn_mp->getView();
+
+        Kokkos::parallel_for(
+            this->A_n.getFieldRangePolicy(), KOKKOS_LAMBDA(size_t i, size_t j, size_t k) {
+                ippl::Vector<scalar, 3> dAdt;
+                dAdt[0] = (A_np1(i, j, k)[1] - A_n(i, j, k)[1]) * idt;
+                dAdt[1] = (A_np1(i, j, k)[2] - A_n(i, j, k)[2]) * idt;
+                dAdt[2] = (A_np1(i, j, k)[3] - A_n(i, j, k)[3]) * idt;
+                // TODO make this work with tail!!!
```
Comment (aaadelmann) [inline src/MaxwellSolvers/FDTDSolverBase.hpp:113]: do not understand the TODO, please describe in more detail. Can mobe this up the the description of the function

File: src/MaxwellSolvers/NonStandardFDTDSolver.hpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
+/**
+ * @file NonStandardFDTDSolver.hpp
+ * @brief Impelmentation of the NonStandardFDTDSolver class functions.
```
Comment (aaadelmann) [inline src/MaxwellSolvers/NonStandardFDTDSolver.hpp:3]: add a citation for the method

File: test/solver/TestNonStandardFDTDSolver.cpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
+#include <cstddef>
```
Comment (aaadelmann) [inline test/solver/TestNonStandardFDTDSolver.cpp:1]: A short description is in order. Maybe just refer to the Read.me

File: test/solver/TestNonStandardFDTDSolver_convergence.cpp [hunk 5]
```diff
@@ -0,0 +1,166 @@
+#include <cstddef>
```
Comment (aaadelmann) [inline test/solver/TestNonStandardFDTDSolver_convergence.cpp:1]: A short description is in order. Maybe just refer to the Read.me

File: test/solver/TestStandardFDTDSolver.cpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
+#include <cstddef>
```
Comment (aaadelmann) [inline test/solver/TestStandardFDTDSolver.cpp:1]: A short description is in order. Maybe just refer to the Read.me

File: test/solver/TestStandardFDTDSolver_convergence.cpp [hunk 7]
```diff
@@ -0,0 +1,166 @@
+#include <cstddef>
```
Comment (aaadelmann) [inline test/solver/TestStandardFDTDSolver_convergence.cpp:1]: A short description is in order. Maybe just refer to the Read.me

**Review by aaadelmann — CHANGES_REQUESTED** [review]

File: src/MaxwellSolvers/FDTDSolverBase.hpp (hunk 2 above)
Comment (aaadelmann) [reply src/MaxwellSolvers/FDTDSolverBase.hpp:113]: agree

**Review by aaadelmann — APPROVED** [review]

File: src/MaxwellSolvers/AbsorbingBC.h [hunk 8]
```diff
@@ -0,0 +1,573 @@
...
+     * @return Updated value at the boundary.
+     */
+    template<typename view_type, typename Coords>
+    KOKKOS_INLINE_FUNCTION auto operator()(const view_type& A_n, const view_type& A_nm1,const view_type& A_np1, const Coords& c)const -> typename view_type::value_type{
+        uint32_t i = c[0];
+        uint32_t j = c[1];
+        uint32_t k = c[2];
+        using ippl::apply;
+        constexpr unsigned side_axes[2] = {_side_axes...};
+        ippl::Vector<uint32_t, 3> side_axis1_onehot = ippl::Vector<uint32_t, 3>{side_axes[0] == 0, side_axes[0] == 1, side_axes[0] == 2};
```
Comment (s-mayani) [inline src/MaxwellSolvers/AbsorbingBC.h:87]: what are these side_axis_onehot ? maybe it would be beneficial to have a comment explaining these in the same way that you explain the offset of the main axis in lines 91-93.

File: src/MaxwellSolvers/AbsorbingBC.h [hunk 9]
```diff
@@ -0,0 +1,573 @@
...
+     */
+    template<typename view_type, typename Coords>
+    KOKKOS_INLINE_FUNCTION auto operator()(const view_type& A_n, const view_type& A_nm1,const view_type& A_np1, const Coords& c)const -> typename view_type::value_type{
+        uint32_t i = c[0];
+        uint32_t j = c[1];
+        uint32_t k = c[2];
+        using ippl::apply;
+
+        // Get the opposite direction of the normal vectors for the two normal axes (direction to the interior of the domain)
+        ippl::Vector<int32_t, 3> normal_axis1_onehot = ippl::Vector<int32_t, 3>{normal_axis1 == 0, normal_axis1 == 1, normal_axis1 == 2} * int32_t(na1_zero ? 1 : -1);
```
Comment (s-mayani) [inline src/MaxwellSolvers/AbsorbingBC.h:190]: maybe same commet as above, what is this _onehot?

File: src/MaxwellSolvers/AbsorbingBC.h [hunk 10]
```diff
@@ -0,0 +1,573 @@
...
+        ippl::Vector<uint32_t, 3> local_nr{
+            uint32_t(A_n.extent(0)),
+            uint32_t(A_n.extent(1)),
+            uint32_t(A_n.extent(2))
+        };
+
+        constexpr uint32_t min_abc_boundary = 1;
+        constexpr uint32_t max_abc_boundary_sub = min_abc_boundary + 1;
+        
+        // Apply second-order ABC to faces
```
Comment (s-mayani) [inline src/MaxwellSolvers/AbsorbingBC.h:397]: why did you decide to have three separate kokkos loops for the face, edge, and corner cases? if there is anyways a check with an if inside, is there a reason to not put all the cases in the same loop? or is it more performant this way?

File: src/MaxwellSolvers/FDTDSolverBase.h [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        
+        SourceField A_n;                            // Current field.
+        SourceField A_np1;                          // Field at the next time step.
+        SourceField A_nm1;                          // Field at the previous time step.
+        
+        /**
+         * @brief Shifts the time for the fields.
+         */
+        void timeShift();
+        protected:
```
Comment (s-mayani) [inline src/MaxwellSolvers/FDTDSolverBase.h:69]: fix indentation - maybe this will be fixed if you apply the IPPL clang formatting

File: src/MaxwellSolvers/FDTDSolverBase.h [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        NDIndex<Dim> domain_m;                      // Domain of the mesh.
+        Vector_t hr_m;                              // Mesh spacing.
+        
+        Vector<int, Dim> nr_m;                      // Number of grid points in each direction.
+        scalar dt;                                  // Time step size.
+
+        /**
+         * @brief Initializes the solver. This is a pure virtual function.
+         */
+        virtual void initialize() = 0;
```
Comment (s-mayani) [inline src/MaxwellSolvers/FDTDSolverBase.h:98]: can you move these to be above the class members? i.e. above line 85

File: src/MaxwellSolvers/FDTDSolverBase.hpp [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+            (void)x;
+        };
+        bcsetter(std::make_index_sequence<Dim * 2>{});
+        A_n.setFieldBC(vector_bcs);
+        A_np1.setFieldBC(vector_bcs);
+        A_nm1.setFieldBC(vector_bcs);
+    }
+
+    /**
+     * @brief Shifts the time for the fields.
```
Comment (s-mayani) [inline src/MaxwellSolvers/FDTDSolverBase.hpp:59]: Maybe I would change this to be "Shifts the saved fields in time"

File: src/MaxwellSolvers/FDTDSolverBase.hpp [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+                ippl::Vector<scalar, 4> dAdx =
+                    (A_n(i + 1, j, k) - A_n(i - 1, j, k)) * inverse_2_spacing[0];
+                ippl::Vector<scalar, 4> dAdy =
+                    (A_n(i, j + 1, k) - A_n(i, j - 1, k)) * inverse_2_spacing[1];
+                ippl::Vector<scalar, 4> dAdz =
+                    (A_n(i, j, k + 1) - A_n(i, j, k - 1)) * inverse_2_spacing[2];
+
+                ippl::Vector<scalar, 3> grad_phi{dAdx[0], dAdy[0], dAdz[0]};
+                ippl::Vector<scalar, 3> curlA{
```
Comment (s-mayani) [inline src/MaxwellSolvers/FDTDSolverBase.hpp:121]: I believe there is an ippl curl operator, maybe you can replace this local implementation of the curl by that?

File: src/MaxwellSolvers/NonStandardFDTDSolver.hpp [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+     * @param source The source field.
+     * @param E The electric field.
+     * @param B The magnetic field.
+     */
+    template <typename EMField, typename SourceField, fdtd_bc boundary_conditions>
+    NonStandardFDTDSolver<EMField, SourceField, boundary_conditions>::NonStandardFDTDSolver(SourceField& source, EMField& E, EMField& B) : FDTDSolverBase<EMField, SourceField, boundary_conditions>(source, E, B) {
+        auto hx = source.get_mesh().getMeshSpacing();
+        if ((hx[2] / hx[0]) * (hx[2] / hx[0]) + (hx[2] / hx[1]) * (hx[2] / hx[1]) >= 1) {
+            std::cerr << "Dispersion-free CFL condition not satisfiable\n";
+            std::abort();
```
Comment (s-mayani) [inline src/MaxwellSolvers/NonStandardFDTDSolver.hpp:23]: should this be replaced by an IPPL exception?

File: test/solver/CMakeLists.txt
Comment (s-mayani) [inline test/solver/CMakeLists.txt:1]: can we move all the FDTD tests to a new solver directory called electromagnetism or maxwell_solver or something like that? what do you think @aaadelmann?

Comment (s-mayani) [reply test/solver/CMakeLists.txt:1]: Having seen the readme below, I think it might make sense to just make a separate directory called fdtd_solver with the tests and put the readme there, as this readme does not concern any of the other tests in test/solver.

File: test/solver/TestNonStandardFDTDSolver.cpp [hunk 16]
```diff
@@ -0,0 +1,164 @@
+// Check the README.md file in this directory for information about the test and how to run it.
```
Comment (s-mayani) [inline test/solver/TestNonStandardFDTDSolver.cpp:1]: I would still have the name of the file at the top here as a header ("TestNonStandardFDTDSolver.cpp")

File: test/solver/TestNonStandardFDTDSolver.cpp [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+        SourceField source(mesh, layout);
+        EMField E(mesh, layout);
+        EMField B(mesh, layout);
+        source = vector4_type(0);
+
+        // Create the NonStandardFDTDSolver object
+        ippl::NonStandardFDTDSolver<EMField, SourceField, ippl::periodic> sfdsolver(source, E, B);
+
+        // Set boundary conditions
+        sfdsolver.setPeriodicBoundaryConditions();
```
Comment (s-mayani) [inline test/solver/TestNonStandardFDTDSolver.cpp:65]: if there is a template parameter "ippl::periodic" being passed to the sfdsolver, why do you need to setPeriodicBoundaryConditions explicitly? can't this be done in the constructor depending on the template parameter?

File: test/solver/TestNonStandardFDTDSolver.cpp [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+            am1view(i,j,k) = vector4_type{scalar(0), scalar(0), magnitude, scalar(0)};
+        });
+        Kokkos::fence();
+
+        // Apply the boundary conditions to the initialized fields
+        sfdsolver.A_n.getFieldBC().apply(sfdsolver.A_n);
+        sfdsolver.A_np1.getFieldBC().apply(sfdsolver.A_np1);
+        sfdsolver.A_nm1.getFieldBC().apply(sfdsolver.A_nm1);
+        sfdsolver.A_n.fillHalo();
+        sfdsolver.A_nm1.fillHalo();
```
Comment (s-mayani) [inline test/solver/TestNonStandardFDTDSolver.cpp:94]: Do we really need these? 

If I look at test/solver/TestFFTPeriodicPoissonSolver.cpp or test/solver/TestCGSolver.cpp, we never need to explicitly do this in the test, but rather we just set the periodic face BCs for the field and then use setRhs to pass the field. 

Also, thinking again about the fields A_n, etc. being public such that they can be accessed like this, can't we just have in the (Non)StandardFDTDSolver class some functions "setA_n", "setA_nm1", "setA_np1" to which you can pass the fields you want to set them with?

File: test/solver/TestNonStandardFDTDSolver_convergence.cpp [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+    SourceField source(mesh, layout);
+    EMField E(mesh, layout);
+    EMField B(mesh, layout);
+    source = vector4_type(0);
+
+    // Create the NonStandardFDTDSolver object
+    ippl::NonStandardFDTDSolver<EMField, SourceField, ippl::periodic> sfdsolver(source, E, B);
+
+    // Set boundary conditions
+    sfdsolver.setPeriodicBoundaryConditions();
```
Comment (s-mayani) [inline test/solver/TestNonStandardFDTDSolver_convergence.cpp:67]: same comments as previous file regarding BCs

File: test/solver/TestNonStandardFDTDSolver_convergence.cpp [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+    });
+    Kokkos::fence();
+
+
+    // Apply the boundary conditions to the initialized fields
+    sfdsolver.A_n.getFieldBC().apply(sfdsolver.A_n);
+    sfdsolver.A_np1.getFieldBC().apply(sfdsolver.A_np1);
+    sfdsolver.A_nm1.getFieldBC().apply(sfdsolver.A_nm1);
+    sfdsolver.A_n.fillHalo();
+    sfdsolver.A_nm1.fillHalo();
```
Comment (s-mayani) [inline test/solver/TestNonStandardFDTDSolver_convergence.cpp:106]: same comments as previous file

File: test/solver/TestNonStandardFDTDSolver_convergence.cpp [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+        aview(i,j,k)[dir] -= gauss(scalar(0.5), scalar(0.05), coord);
+        ref += Kokkos::sqrt(aview(i,j,k)[0] * aview(i,j,k)[0] + aview(i,j,k)[1] * aview(i,j,k)[1] + aview(i,j,k)[2] * aview(i,j,k)[2]);
+    }, sum_error);
+
+    // Write results to output file
+    Inform csvout(NULL, fname.c_str(), Inform::APPEND);
+    csvout.precision(16);
+    csvout.setf(std::ios::scientific, std::ios::floatfield);
+
+    csvout << direction << "," << np << "," << sum_error / double(nr[1] * nr[1] * nr[2]) << endl;
```
Comment (s-mayani) [inline test/solver/TestNonStandardFDTDSolver_convergence.cpp:141]: not sure this is the correct computation of the error, see my previous comment!

File: test/solver/TestStandardFDTDSolver.cpp [hunk 22]
```diff
@@ -0,0 +1,164 @@
+// Check the README.md file in this directory for information about the test and how to run it.
```
Comment (s-mayani) [inline test/solver/TestStandardFDTDSolver.cpp:1]: add file name at top

File: test/solver/TestStandardFDTDSolver.cpp [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+    uint32_t dim = sizeof...(scalar);
+    ippl::Vector<scalar1, sizeof...(scalar)> vec{scalar1(x - mean)...};
+    scalar1 vecsum(0);
+    for(unsigned d = 0;d < dim;d++){
+        vecsum += vec[d] * vec[d];
+        
+    }
+    #ifndef __CUDA_ARCH__
+    using std::exp;
+    #endif
```
Comment (s-mayani) [inline test/solver/TestStandardFDTDSolver.cpp:24]: same comment as before

File: test/solver/TestStandardFDTDSolver.cpp [hunk 24]
```diff
@@ -0,0 +1,164 @@
...
+        SourceField source(mesh, layout);
+        EMField E(mesh, layout);
+        EMField B(mesh, layout);
+        source = vector4_type(0);
+
+        // Create the StandardFDTDSolver object
+        ippl::StandardFDTDSolver<EMField, SourceField, ippl::periodic> sfdsolver(source, E, B);
+
+        // Set boundary conditions
+        sfdsolver.setPeriodicBoundaryConditions();
```
Comment (s-mayani) [inline test/solver/TestStandardFDTDSolver.cpp:65]: same as before

File: test/solver/TestStandardFDTDSolver.cpp [hunk 25]
```diff
@@ -0,0 +1,164 @@
...
+            am1view(i,j,k) = vector4_type{scalar(0), scalar(0), magnitude, scalar(0)};
+        });
+        Kokkos::fence();
+
+        // Apply the boundary conditions to the initialized fields
+        sfdsolver.A_n.getFieldBC().apply(sfdsolver.A_n);
+        sfdsolver.A_np1.getFieldBC().apply(sfdsolver.A_np1);
+        sfdsolver.A_nm1.getFieldBC().apply(sfdsolver.A_nm1);
+        sfdsolver.A_n.fillHalo();
+        sfdsolver.A_nm1.fillHalo();
```
Comment (s-mayani) [inline test/solver/TestStandardFDTDSolver.cpp:94]: same as before

File: test/solver/TestStandardFDTDSolver.cpp [hunk 26]
```diff
@@ -0,0 +1,164 @@
...
+            scalar x = scalar(ig) / nr[0];
+            scalar y = scalar(jg) / nr[1];
+            scalar z = scalar(kg) / nr[2];
+            (void)x;
+            (void)y;
+            (void)z;
+            ref += Kokkos::abs(gauss(scalar(0.5), scalar(0.05), z) - aview(i,j,k)[2]);
+        }, sum_error);
+        std::cout << "Sum after evaluating boundaryconditions: " << sum_error / double(nr[0] * nr[1] * nr[2]) << "\n";
+
```
Comment (s-mayani) [inline test/solver/TestStandardFDTDSolver.cpp:161]: check error computation

File: test/solver/TestStandardFDTDSolver_convergence.cpp [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+    uint32_t dim = sizeof...(scalar);
+    ippl::Vector<scalar1, sizeof...(scalar)> vec{scalar1(x - mean)...};
+    scalar1 vecsum(0);
+    for(unsigned d = 0;d < dim;d++){
+        vecsum += vec[d] * vec[d];
+        
+    }
+    #ifndef __CUDA_ARCH__
+    using std::exp;
+    #endif
```
Comment (s-mayani) [inline test/solver/TestStandardFDTDSolver_convergence.cpp:23]: same as before

File: test/solver/TestStandardFDTDSolver_convergence.cpp [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+    EMField E(mesh, layout);
+    EMField B(mesh, layout);
+    source = vector4_type(0);
+
+    // Create the StandardFDTDSolver object
+    ippl::StandardFDTDSolver<EMField, SourceField, ippl::periodic> sfdsolver(source, E, B);
+
+    // Set boundary conditions
+    sfdsolver.setPeriodicBoundaryConditions();
+
```
Comment (s-mayani) [inline test/solver/TestStandardFDTDSolver_convergence.cpp:68]: same as before

File: test/solver/TestStandardFDTDSolver_convergence.cpp [hunk 29]
```diff
@@ -0,0 +1,168 @@
...
+        aview(i,j,k)[dir] -= gauss(scalar(0.5), scalar(0.05), coord);
+        ref += Kokkos::sqrt(aview(i,j,k)[0] * aview(i,j,k)[0] + aview(i,j,k)[1] * aview(i,j,k)[1] + aview(i,j,k)[2] * aview(i,j,k)[2]);
+    }, sum_error);
+
+    // Write results to output file
+    Inform csvout(NULL, fname.c_str(), Inform::APPEND);
+    csvout.precision(16);
+    csvout.setf(std::ios::scientific, std::ios::floatfield);
+
+    csvout << direction << "," << np << "," << sum_error / double(nr[1] * nr[1] * nr[2]) << endl;
```
Comment (s-mayani) [inline test/solver/TestStandardFDTDSolver_convergence.cpp:141]: check error calculation

**Review by s-mayani — CHANGES_REQUESTED** [review]
Left some comments.
In general, I think the IPPL clang formatting needs to be applied.
Also, I am not sure about the error computation in the tests.

File: src/MaxwellSolvers/AbsorbingBC.h (hunk 10 above)
Comment (anekslen) [reply src/MaxwellSolvers/AbsorbingBC.h:397]: This is probably performance related, I am not sure tough.

Since we have a return statement in the first two loops, directly after calculating lval (line 408 and 460), we should not loose a lot of time, I would leave it the way it is.

File: src/MaxwellSolvers/FDTDSolverBase.hpp (hunk 14 above)
Comment (anekslen) [reply src/MaxwellSolvers/FDTDSolverBase.hpp:121]: This is done, because dAdx is a vector with 4 dimensions and we only want the curl over the last three dimensions.

So using the standard ippl::curl is not possible

Comment (s-mayani) [reply src/MaxwellSolvers/FDTDSolverBase.hpp:121]: Ah okay, thanks for the answer, maybe put a comment above this part of the code explaining this

File: test/solver/TestNonStandardFDTDSolver.cpp (hunk 17 above)
Comment (anekslen) [reply test/solver/TestNonStandardFDTDSolver.cpp:65]: This is done in the constructor of the Solver now.

File: test/solver/TestNonStandardFDTDSolver.cpp [hunk 30]
```diff
@@ -0,0 +1,164 @@
...
+    uint32_t dim = sizeof...(scalar);
+    ippl::Vector<scalar1, sizeof...(scalar)> vec{scalar1(x - mean)...};
+    scalar1 vecsum(0);
+    for(unsigned d = 0;d < dim;d++){
+        vecsum += vec[d] * vec[d];
+        
+    }
+    #ifndef __CUDA_ARCH__
+    using std::exp;
+    #endif
```
[the remark this answers is not in the document]
Comment (s-mayani) [reply test/solver/TestNonStandardFDTDSolver.cpp:24]: alright sounds good, then I would say remove this and replace the below call to exp() by Kokkos::exp(), in all the places where this appears

File: src/MaxwellSolvers/FDTDSolverBase.h [hunk 31]
```diff
@@ -0,0 +1,102 @@
...
+        virtual void step() = 0;
+        /**
+         * @brief Evaluates the electric and magnetic fields.
+         */
+        void evaluate_EB();
+
+        /**
+         * @brief Initializes the solver. This is a pure virtual function.
+         */
+        virtual void initialize() = 0;
```
Comment (s-mayani) [inline src/MaxwellSolvers/FDTDSolverBase.h:89]: can these also be moved to be with the public members above the protected so they are in the same place? thanks

File: src/MaxwellSolvers/FDTDSolverBase.h [hunk 32]
```diff
@@ -0,0 +1,102 @@
...
+
+        typename SourceField::Mesh_t* mesh_mp;  // Pointer to the mesh.
+        FieldLayout<Dim>* layout_mp;            // Pointer to the layout
+        NDIndex<Dim> domain_m;                  // Domain of the mesh.
+        Vector_t hr_m;                          // Mesh spacing.
+
+        Vector<int, Dim> nr_m;  // Number of grid points in each direction.
+        scalar dt;              // Time step size.
```
Comment (s-mayani) [inline src/MaxwellSolvers/FDTDSolverBase.h:97]: why do we need these class members to be public? could these be private?
